integration/spark/src/test/common/java/io/openlineage/spark/agent/SparkOpenLineageFailuresTest.java

```diff
@@ -0,0 +1,148 @@
+
+  private static final Network network = Network.newNetwork();
+
+  @Container
+  private static final MockServerContainer openLineageClientMockContainer =
+      SparkContainerUtils.makeMockServerContainer(network);
+
+  private static final String INCORRECT_PORT = "3000";
+  private static final String CORRECT_PORT = "1080";
+
+  private static MockServerClient mockServerClient;
+
+  @AfterEach
+  public void cleanupSpark() {
+    if (mockServerClient != null) {
+      mockServerClient.reset();
+    }
+  }
+
+  @Test
+  public void testIncorrectOpenLineageEndpointUrl() {
+    SparkContainerUtils.makePysparkContainerWithDefaultConf(
+            network,
+            "http://openlineageclient:" + INCORRECT_PORT,
+            ".*Spark is fine!.*",
+            openLineageClientMockContainer,
+            "testFailures",
+            "/opt/spark_scripts/spark_test_failures.py")
+        .start();
+  }
+
+  @Test
+  public void testOpenLineageEndpointResponds400() {
+    configureMockServerToRespondWith(400);
+
+    SparkContainerUtils.makePysparkContainerWithDefaultConf(
+            network,
+            "http://openlineageclient:" + CORRECT_PORT,
+            ".*Spark is fine!.*",
+            openLineageClientMockContainer,
+            "testFailures",
+            "/opt/spark_scripts/spark_test_failures.py")
+        .start();
+  }
+
+  @Test
+  public void testOpenLineageEndpointResponds500() {
+    configureMockServerToRespondWith(500);
+
+    SparkContainerUtils.makePysparkContainerWithDefaultConf(
+            network,
+            "http://openlineageclient:" + CORRECT_PORT,
+            ".*Spark is fine!.*",
+            openLineageClientMockContainer,
+            "testFailures",
+            "/opt/spark_scripts/spark_test_failures.py")
+        .start();
+  }
+
+  @Test
+  public void testOpenLineageFailingVisitor(SparkSession spark)
+      throws InterruptedException, TimeoutException {
+    StructType schema =
+        new StructType(
+            new StructField[] {
+              new StructField("aString", StringType$.MODULE$, false, new Metadata(new HashMap<>()))
+            });
+
+    // make sure table location in warehouse dir is empty
+    cleanLocation(spark, FAILING_TABLE_NAME_FAIL_ON_APPLY);
+    cleanLocation(spark, FAILING_TABLE_NAME_FAIL_ON_IS_DEFINED);
+
+    Dataset<Row> dataset =
+        spark.createDataFrame(Arrays.asList(new GenericRow(new Object[] {"one"})), schema);
+
+    dataset.write().mode(SaveMode.Overwrite).saveAsTable(FAILING_TABLE_NAME_FAIL_ON_IS_DEFINED);
+    dataset.write().mode(SaveMode.Overwrite).saveAsTable(FAILING_TABLE_NAME_FAIL_ON_APPLY);
+    // creating table that way should fail because of a failing visitor defined in
+    // TestOpenLineageEventHandlerFactory
+
+    StaticExecutionContextFactory.waitForExecutionEnd();
+    assertEquals(1, spark.table(FAILING_TABLE_NAME_FAIL_ON_IS_DEFINED).count());
+    assertEquals(1, spark.table(FAILING_TABLE_NAME_FAIL_ON_APPLY).count());
```

Comment:
Can we add one more assertion on the message sent to the backend (if any)?

Reply:
ok, good point.